test/integration/health_check_integration_test.cc

```diff
@@ -0,0 +1,261 @@
+    registerTestServerPorts({"http"});
+
+    // Create the regular (i.e. not an xDS server) upstreams. We create them manually here after
+    // initialize() because finalize() expects all fake_upstreams_ to correspond to a static
+    // cluster in the bootstrap config - which we don't want since we're using dynamic CDS.
+    const FakeHttpConnection::Type http_conn_type =
+        (envoy::type::v3::CodecClientType::HTTP1 == codec_client_type_)
+            ? FakeHttpConnection::Type::HTTP1
+            : FakeHttpConnection::Type::HTTP2;
+    for (auto& cluster : clusters_) {
+      cluster.host_upstream_ = std::make_unique<FakeUpstream>(0, http_conn_type, version_,
+                                                              timeSystem(), enable_half_close_);
+      cluster.cluster_ = ConfigHelper::buildStaticCluster(
+          cluster.name_, cluster.host_upstream_->localAddress()->ip()->port(),
+          Network::Test::getLoopbackAddressString(GetParam()));
+    }
+  }
+
+  void acceptXdsConnection() {
+    AssertionResult result = // xds_connection_ is filled with the new FakeHttpConnection.
+        fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, xds_connection_);
+    RELEASE_ASSERT(result, result.message());
+    result = xds_connection_->waitForNewStream(*dispatcher_, xds_stream_);
+    RELEASE_ASSERT(result, result.message());
+    xds_stream_->startGrpcStream();
+  }
+
+  // Closes the connections to the fake hosts.
+  void cleanupHostConnections() {
+    for (auto& cluster : clusters_) {
+      auto& host_fake_connection = cluster.host_fake_connection_;
+      if (host_fake_connection != nullptr) {
+        AssertionResult result = host_fake_connection->close();
+        RELEASE_ASSERT(result, result.message());
+        result = host_fake_connection->waitForDisconnect();
+        RELEASE_ASSERT(result, result.message());
+      }
+    }
+  }
+
+  // Adds an active health check specifier to the given cluster with the given timeout.
+  envoy::config::core::v3::HealthCheck* addHealthCheck(envoy::config::cluster::v3::Cluster& cluster,
+                                                       uint32_t timeout) {
+    // Add general health check specifier to the cluster.
+    auto* health_check = cluster.add_health_checks();
+    health_check->mutable_timeout()->set_seconds(timeout);
+    health_check->mutable_interval()->CopyFrom(
+        Protobuf::util::TimeUtil::MillisecondsToDuration(100));
+    health_check->mutable_no_traffic_interval()->CopyFrom(
+        Protobuf::util::TimeUtil::MillisecondsToDuration(100));
+    health_check->mutable_unhealthy_threshold()->set_value(1);
+    health_check->mutable_healthy_threshold()->set_value(1);
+    return health_check;
+  }
+
+  // Adds a HTTP active health check specifier to the given cluster with the given timeout,
+  // and waits for the first health check probe to be received.
+  void initHttpHealthCheck(uint32_t cluster_idx, uint32_t timeout = 30) {
```

Comment:
is this 30 second timeout a real time? if so is it possible to use simulated time in this class?

Reply:
Changed to using simulated time.
Thanks for suggesting it.